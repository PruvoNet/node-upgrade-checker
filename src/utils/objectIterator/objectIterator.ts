export interface IRootNode {
  value: any;
}

export interface INode extends IRootNode {
  parent: IParent;
  key: string;
  depth: number;
}

export type IParent = INode | IRootNode;
export type ObjectIterator = Generator<INode, void, boolean | undefined>;

export function* objectIterator(obj: any): ObjectIterator {
  yield* iterator({ value: obj }, obj, 0);
}

function* iterator(parent: IParent, obj: any, depth: number): ObjectIterator {
  if (!isObject(obj)) {
    return;
  }
  const nextDepth = depth + 1;
  for (const key of Object.keys(obj)) {
    const value = obj[key];
    const node: INode = { key, value, depth, parent };
    if (!(yield node)) {
      yield* iterator(node, value, nextDepth);
    }
  }
}

const isObject = (obj: any): obj is Record<string, any> => {
  return obj !== null && typeof obj === `object`;
};
