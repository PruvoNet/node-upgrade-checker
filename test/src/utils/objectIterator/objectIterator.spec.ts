import { INode, objectIterator } from '../../../../src/utils/objectIterator/objectIterator';

describe(`object iterator`, () => {
  describe(`for of`, () => {
    it(`Should iterate simple object`, () => {
      const obj = {
        a: 1,
        b: {
          c: 2,
        },
        d: `7`,
      };

      const results: INode[] = [];
      for (const node of objectIterator(obj)) {
        results.push(node);
      }
      expect(results).toEqual([
        {
          key: `a`,
          value: 1,
          depth: 0,
          parent: { value: { a: 1, b: { c: 2 }, d: `7` } },
        },
        {
          key: `b`,
          value: { c: 2 },
          depth: 0,
          parent: { value: { a: 1, b: { c: 2 }, d: `7` } },
        },
        {
          key: `c`,
          value: 2,
          depth: 1,
          parent: { key: `b`, value: { c: 2 }, depth: 0, parent: { value: { a: 1, b: { c: 2 }, d: `7` } } },
        },
        { key: `d`, value: `7`, depth: 0, parent: { value: { a: 1, b: { c: 2 }, d: `7` } } },
      ]);
    });

    it(`Should iterate complex object`, () => {
      const obj = {
        a: 1,
        b: {
          c: 2,
          d: {
            e: `5`,
            f: [1, `26`],
          },
        },
        g: `7`,
      };

      const results: INode[] = [];
      for (const node of objectIterator(obj)) {
        results.push(node);
      }
      expect(results).toMatchSnapshot();
    });
  });
  describe(`skip logic`, () => {
    it(`Should iterate simple object`, () => {
      const obj = {
        a: 1,
        b: {
          c: 2,
        },
        d: `7`,
      };
      const results: INode[] = [];
      const it = objectIterator(obj);
      let result = it.next();
      while (!result.done) {
        const { value: node } = result;
        results.push(node);
        result = it.next(node.key === `b`);
      }
      expect(results).toEqual([
        {
          key: `a`,
          value: 1,
          depth: 0,
          parent: { value: { a: 1, b: { c: 2 }, d: `7` } },
        },
        {
          key: `b`,
          value: { c: 2 },
          depth: 0,
          parent: { value: { a: 1, b: { c: 2 }, d: `7` } },
        },
        { key: `d`, value: `7`, depth: 0, parent: { value: { a: 1, b: { c: 2 }, d: `7` } } },
      ]);
    });

    it(`Should iterate complex object`, () => {
      const obj = {
        a: 1,
        b: {
          c: 2,
          d: {
            e: `5`,
            f: [1, `26`],
          },
        },
        g: `7`,
      };
      const results: INode[] = [];
      const it = objectIterator(obj);
      let result = it.next();
      while (!result.done) {
        const { value: node } = result;
        results.push(node);
        result = it.next(node.key === `d`);
      }
      expect(results).toMatchSnapshot();
    });
  });
});
