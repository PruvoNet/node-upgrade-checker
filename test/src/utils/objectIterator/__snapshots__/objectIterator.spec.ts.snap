// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`object iterator for of Should iterate complex object 1`] = `
Array [
  Object {
    "depth": 0,
    "key": "a",
    "parent": Object {
      "value": Object {
        "a": 1,
        "b": Object {
          "c": 2,
          "d": Object {
            "e": "5",
            "f": Array [
              1,
              "26",
            ],
          },
        },
        "g": "7",
      },
    },
    "value": 1,
  },
  Object {
    "depth": 0,
    "key": "b",
    "parent": Object {
      "value": Object {
        "a": 1,
        "b": Object {
          "c": 2,
          "d": Object {
            "e": "5",
            "f": Array [
              1,
              "26",
            ],
          },
        },
        "g": "7",
      },
    },
    "value": Object {
      "c": 2,
      "d": Object {
        "e": "5",
        "f": Array [
          1,
          "26",
        ],
      },
    },
  },
  Object {
    "depth": 1,
    "key": "c",
    "parent": Object {
      "depth": 0,
      "key": "b",
      "parent": Object {
        "value": Object {
          "a": 1,
          "b": Object {
            "c": 2,
            "d": Object {
              "e": "5",
              "f": Array [
                1,
                "26",
              ],
            },
          },
          "g": "7",
        },
      },
      "value": Object {
        "c": 2,
        "d": Object {
          "e": "5",
          "f": Array [
            1,
            "26",
          ],
        },
      },
    },
    "value": 2,
  },
  Object {
    "depth": 1,
    "key": "d",
    "parent": Object {
      "depth": 0,
      "key": "b",
      "parent": Object {
        "value": Object {
          "a": 1,
          "b": Object {
            "c": 2,
            "d": Object {
              "e": "5",
              "f": Array [
                1,
                "26",
              ],
            },
          },
          "g": "7",
        },
      },
      "value": Object {
        "c": 2,
        "d": Object {
          "e": "5",
          "f": Array [
            1,
            "26",
          ],
        },
      },
    },
    "value": Object {
      "e": "5",
      "f": Array [
        1,
        "26",
      ],
    },
  },
  Object {
    "depth": 2,
    "key": "e",
    "parent": Object {
      "depth": 1,
      "key": "d",
      "parent": Object {
        "depth": 0,
        "key": "b",
        "parent": Object {
          "value": Object {
            "a": 1,
            "b": Object {
              "c": 2,
              "d": Object {
                "e": "5",
                "f": Array [
                  1,
                  "26",
                ],
              },
            },
            "g": "7",
          },
        },
        "value": Object {
          "c": 2,
          "d": Object {
            "e": "5",
            "f": Array [
              1,
              "26",
            ],
          },
        },
      },
      "value": Object {
        "e": "5",
        "f": Array [
          1,
          "26",
        ],
      },
    },
    "value": "5",
  },
  Object {
    "depth": 2,
    "key": "f",
    "parent": Object {
      "depth": 1,
      "key": "d",
      "parent": Object {
        "depth": 0,
        "key": "b",
        "parent": Object {
          "value": Object {
            "a": 1,
            "b": Object {
              "c": 2,
              "d": Object {
                "e": "5",
                "f": Array [
                  1,
                  "26",
                ],
              },
            },
            "g": "7",
          },
        },
        "value": Object {
          "c": 2,
          "d": Object {
            "e": "5",
            "f": Array [
              1,
              "26",
            ],
          },
        },
      },
      "value": Object {
        "e": "5",
        "f": Array [
          1,
          "26",
        ],
      },
    },
    "value": Array [
      1,
      "26",
    ],
  },
  Object {
    "depth": 3,
    "key": "0",
    "parent": Object {
      "depth": 2,
      "key": "f",
      "parent": Object {
        "depth": 1,
        "key": "d",
        "parent": Object {
          "depth": 0,
          "key": "b",
          "parent": Object {
            "value": Object {
              "a": 1,
              "b": Object {
                "c": 2,
                "d": Object {
                  "e": "5",
                  "f": Array [
                    1,
                    "26",
                  ],
                },
              },
              "g": "7",
            },
          },
          "value": Object {
            "c": 2,
            "d": Object {
              "e": "5",
              "f": Array [
                1,
                "26",
              ],
            },
          },
        },
        "value": Object {
          "e": "5",
          "f": Array [
            1,
            "26",
          ],
        },
      },
      "value": Array [
        1,
        "26",
      ],
    },
    "value": 1,
  },
  Object {
    "depth": 3,
    "key": "1",
    "parent": Object {
      "depth": 2,
      "key": "f",
      "parent": Object {
        "depth": 1,
        "key": "d",
        "parent": Object {
          "depth": 0,
          "key": "b",
          "parent": Object {
            "value": Object {
              "a": 1,
              "b": Object {
                "c": 2,
                "d": Object {
                  "e": "5",
                  "f": Array [
                    1,
                    "26",
                  ],
                },
              },
              "g": "7",
            },
          },
          "value": Object {
            "c": 2,
            "d": Object {
              "e": "5",
              "f": Array [
                1,
                "26",
              ],
            },
          },
        },
        "value": Object {
          "e": "5",
          "f": Array [
            1,
            "26",
          ],
        },
      },
      "value": Array [
        1,
        "26",
      ],
    },
    "value": "26",
  },
  Object {
    "depth": 0,
    "key": "g",
    "parent": Object {
      "value": Object {
        "a": 1,
        "b": Object {
          "c": 2,
          "d": Object {
            "e": "5",
            "f": Array [
              1,
              "26",
            ],
          },
        },
        "g": "7",
      },
    },
    "value": "7",
  },
]
`;

exports[`object iterator skip logic Should iterate complex object 1`] = `
Array [
  Object {
    "depth": 0,
    "key": "a",
    "parent": Object {
      "value": Object {
        "a": 1,
        "b": Object {
          "c": 2,
          "d": Object {
            "e": "5",
            "f": Array [
              1,
              "26",
            ],
          },
        },
        "g": "7",
      },
    },
    "value": 1,
  },
  Object {
    "depth": 0,
    "key": "b",
    "parent": Object {
      "value": Object {
        "a": 1,
        "b": Object {
          "c": 2,
          "d": Object {
            "e": "5",
            "f": Array [
              1,
              "26",
            ],
          },
        },
        "g": "7",
      },
    },
    "value": Object {
      "c": 2,
      "d": Object {
        "e": "5",
        "f": Array [
          1,
          "26",
        ],
      },
    },
  },
  Object {
    "depth": 1,
    "key": "c",
    "parent": Object {
      "depth": 0,
      "key": "b",
      "parent": Object {
        "value": Object {
          "a": 1,
          "b": Object {
            "c": 2,
            "d": Object {
              "e": "5",
              "f": Array [
                1,
                "26",
              ],
            },
          },
          "g": "7",
        },
      },
      "value": Object {
        "c": 2,
        "d": Object {
          "e": "5",
          "f": Array [
            1,
            "26",
          ],
        },
      },
    },
    "value": 2,
  },
  Object {
    "depth": 1,
    "key": "d",
    "parent": Object {
      "depth": 0,
      "key": "b",
      "parent": Object {
        "value": Object {
          "a": 1,
          "b": Object {
            "c": 2,
            "d": Object {
              "e": "5",
              "f": Array [
                1,
                "26",
              ],
            },
          },
          "g": "7",
        },
      },
      "value": Object {
        "c": 2,
        "d": Object {
          "e": "5",
          "f": Array [
            1,
            "26",
          ],
        },
      },
    },
    "value": Object {
      "e": "5",
      "f": Array [
        1,
        "26",
      ],
    },
  },
  Object {
    "depth": 0,
    "key": "g",
    "parent": Object {
      "value": Object {
        "a": 1,
        "b": Object {
          "c": 2,
          "d": Object {
            "e": "5",
            "f": Array [
              1,
              "26",
            ],
          },
        },
        "g": "7",
      },
    },
    "value": "7",
  },
]
`;
